### PR TITLE
feat: add volume idle duration cleanup feature (#2497)

### PR DESCRIPTION
**Pull request by haydnli-shopify**
# Summary

Adds automatic cleanup of unused volumes after a configurable idle duration. Users can now set `auto_cleanup_duration` for volumes; if a volume isn't used by any job for that period, it's deleted automatically.

Addresses [Feature]: Delete unused volumes after idle duration #2497.

# Changes

- Add `auto_cleanup_duration` to `VolumeConfiguration`
- Track last usage with `last_job_processed_at` in `VolumeModel`
- Background task (`process_idle_volumes`) runs every 60s to delete idle volumes
- Update job processing to refresh volume usage timestamps
- Database migration for new field
- Unit and integration tests for all scenarios
- Validation to prevent external volumes from using auto-cleanup (not supported)

# Usage

- Set `auto_cleanup_duration` in volume config (YAML or API)
- Volumes not used for the specified time are auto-deleted
- Volumes still attached or with `auto_cleanup_duration: off` are never deleted
- External volumes (with `volume_id`) cannot use auto-cleanup

# Testing

- All core logic and edge cases are covered by tests
- Validation tests ensure external volumes are properly handled

### CODE REVIEW

File: src/dstack/_internal/core/models/volumes.py [hunk 1]
```diff
@@ -44,6 +45,16 @@ class VolumeConfiguration(CoreModel):
         Optional[str],
         Field(description="The volume ID. Must be specified when registering external volumes"),
     ] = None
+    idle_duration: Annotated[
+        Optional[Union[str, int, bool]],
```
Comment (dakota-shop) [inline src/dstack/_internal/core/models/volumes.py]: Why do we need `int`? If 'off' means disable, why do we need `bool`?

File: src/dstack/_internal/core/models/volumes.py [hunk 2]
```diff
@@ -44,6 +45,16 @@ class VolumeConfiguration(CoreModel):
         Optional[str],
         Field(description="The volume ID. Must be specified when registering external volumes"),
     ] = None
+    idle_duration: Annotated[
```
Comment (dakota-shop) [inline src/dstack/_internal/core/models/volumes.py]: Should this be named `max_idle_duration`, or maybe `auto_cleanup_idle_duration`?

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        if project not in volumes_by_project:
+            volumes_by_project[project] = []
+        volumes_by_project[project].append(volume_model.name)
+
+    for project, volume_names in volumes_by_project.items():
+        logger.info("Deleting idle volumes for project %s: %s", project.name, volume_names)
+        try:
+            await delete_volumes(session, project, volume_names)
+        except Exception as e:
+            logger.error("Failed to delete idle volumes for project %s: %s", project.name, str(e))
```
Comment (dakota-shop) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: If you want to log the stacktrace here (looks like we do), use `logger.exception` and leave out `str(e)` (assuming the dstack logger is a Python-standard logging interface)

File: src/dstack/_internal/core/models/volumes.py (hunk 1 above)
Comment (haydnli-shopify) [reply src/dstack/_internal/core/models/volumes.py]: Yes, the boolean is no longer needed here, we could use string 'off' to overlap this
However, I would suggest keeping integer here, 
     since the integer here could be use for specifying idle duration in seconds directly, and -1 as a special value for specifying disabled

Comment (dakota-shop) [reply src/dstack/_internal/core/models/volumes.py]: Why would it be useful to specify this value in seconds when we can force it to be supplied as a human-readable string?

File: src/dstack/_internal/core/models/volumes.py (hunk 2 above)
Comment (haydnli-shopify) [reply src/dstack/_internal/core/models/volumes.py]: Yes, I will have all the name changed!

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py (hunk 3 above)
Comment (haydnli-shopify) [reply src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Sure, I have replace all the logger function with the correct ones

**Comment by haydnli-shopify** [comment]
For this PR,
I've tested it locally and it's working correctly. I created test volumes with different auto-cleanup settings (30s, disabled, integer values) and verified the configuration parsing works. I tested the full lifecycle - volumes get created, go idle, and are automatically deleted after the specified duration. The background task handles all the edge cases we discussed.

If needed, I can upload my test configurations to show the testing approach.

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                .order_by(VolumeModel.last_processed_at.asc())
+                .limit(10)
+                .with_for_update(skip_locked=True)
+            )
+            volumes = list(res.unique().scalars().all())
+            if not volumes:
+                return
+            for volume in volumes:
+                await session.refresh(volume, ["project", "attachments"])
+                lockset.add(volume.id)
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Should not issue multiple queries under `async with lock` since this lock is supposed to be released asap. Instead take lock, select ids, add them to lockset, release the lock, then load the volumes with related attributes in one query – see how other processing tasks do that.

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    by_project = {}
+    for volume in volumes:
+        project = volume.project
+        if project not in by_project:
+            by_project[project] = []
+        by_project[project].append(volume.name)
+
+    for project, names in by_project.items():
+        try:
+            await delete_volumes(session, project, names)
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: delete_volumes does the locking and it's not reentrant so you cannot use it from a task that locks too. It also commits halfway through, so your select for update will end there. Better to write a new deleting function in process_idle_volumes.

File: src/tests/_internal/server/background/tasks/test_process_idle_volumes.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            volume_provisioning_data=get_volume_provisioning_data(),
+        )
+
+        volume.last_job_processed_at = (
+            datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(hours=2)
+        ).replace(tzinfo=None)
+        await session.commit()
+
+        with patch(
+            "dstack._internal.server.background.tasks.process_idle_volumes.delete_volumes"
```
Comment (r4victor) [inline src/tests/_internal/server/background/tasks/test_process_idle_volumes.py]: Why mocking?

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py (hunk 4 above)
Comment (haydnli-shopify) [reply src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Thanks for the response! I've updated it to take the lock and only select volume IDs, add them to the lockset and release the lock immediately, then load the full volumes with related attributes in a single query outside the lock. 
In this way our lock is released as soon as possible, and we still get all the data we need efficiently.

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py (hunk 5 above)
Comment (haydnli-shopify) [reply src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: I've fixed this by creating a new _delete_idle_volumes function that handles the deletion directly
this function simply marks volumes as deleted in the database, i think this could avoids the reentrant lockin here..

File: src/tests/_internal/server/background/tasks/test_process_idle_volumes.py (hunk 6 above)
Comment (haydnli-shopify) [reply src/tests/_internal/server/background/tasks/test_process_idle_volumes.py]: I've removed the mocking and updated the test
the previous commit was mocking delete_volumes and just checking that it was called, but I realized tht doesn't actually test if the volume gets deleted properly. 
the current test runs the background task and then verifies that the volume is actually marked as deleted in the database by checking volume.deleted and volume.deleted_at.

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            for volume in volumes:
+                if _should_delete_volume(volume):
+                    to_delete.append(volume)
+
+            if to_delete:
+                await _delete_idle_volumes(session, to_delete)
+
+        finally:
+            for volume_id in volume_ids:
+                lockset.discard(volume_id)
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Can use lockset.difference_update

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+from dstack._internal.utils.common import get_current_datetime
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+async def process_idle_volumes():
+    lock, lockset = get_locker(get_db().dialect_name).get_lockset(VolumeModel.__tablename__)
+    async with get_session_ctx() as session:
+        # Take lock, select IDs, add to lockset, release lock
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]:
```suggestion
```

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    last_used = volume.last_job_processed_at or volume.created_at
+    last_used_utc = last_used.replace(tzinfo=datetime.timezone.utc)
+    now = get_current_datetime()
+
+    idle_time = now - last_used_utc
+    return max(idle_time, datetime.timedelta(0))
+
+
+async def _delete_idle_volumes(session: AsyncSession, volumes: List[VolumeModel]):
+    """Delete idle volumes without using the delete_volumes function to avoid locking conflicts."""
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Not a useful, confusing docstring

File: ~/.dstack/server/config.yml
Comment (r4victor) [inline ~/.dstack/server/config.yml:1]: Accidentally committed?

**Review by r4victor — CHANGES_REQUESTED** [review]
The PR is ok but it does not actually deletes idle volumes – just marks them as deleted in the db.

**Comment by r4victor** [comment]
The issue and the PR description say to support `idle_duration` for volumes, but the PR actually supports `auto_cleanup_duration`. Which one have you intended to support?

File: src/dstack/_internal/server/background/tasks/process_idle_volumes.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    await session.commit()
+
+
+async def _delete_volume_from_cloud(session: AsyncSession, volume_model: VolumeModel):
+    """Delete volume from cloud provider. Based on volumes.py:_delete_volume"""
+    volume = volume_model_to_volume(volume_model)
+
+    if volume.external:
+        return
```
Comment (r4victor) [inline src/dstack/_internal/server/background/tasks/process_idle_volumes.py]: Maybe disallow creating external volumes with auto_cleanup_duration if it's not supported?

**Review by r4victor — CHANGES_REQUESTED** [review]
Created a volume:

```
type: volume
name: my-gcp-volume
backend: gcp
region: europe-west4
availability_zone: europe-west4-b
size: 100GB
auto_cleanup_duration: 1m
```

process_idle_volumes fails with exception `sqlalchemy.exc.MissingGreenlet: greenlet_spawn` – should be easy to reproduce if creating any volume.

File: src/dstack/_internal/server/migrations/versions/d268739bd76b_merge_volume_cleanup_and_secrets_.py
Comment (r4victor) [inline src/dstack/_internal/server/migrations/versions/d268739bd76b_merge_volume_cleanup_and_secrets_.py:1]: Avoid merge db migrations, rebase migrations on master

**Comment by haydnli-shopify** [comment]
@r4victor Thanks for your response, I have fixed the issue you mentioned, and the PR description has been updated! A force push happened in this fix because I removed the merge migration as requested, updated the migration chain to properly rebase on master, and added validation to prevent external volumes from using auto_cleanup_duration at creation time.

**Comment by r4victor** [comment]
@haydnli-shopify, Just do a regular merge to pull changes from master to avoid force push. No need to rebase the branch on master since we'll merge with squash anyway.

**Comment by haydnli-shopify** [comment]
> @haydnli-shopify, Just do a regular merge to pull changes from master to avoid force push. No need to rebase the branch on master since we'll merge with squash anyway.

Sure! Thanks for your response. Feel free to let me know if there's anything to be improved in this PR!

**Review by r4victor — APPROVED** [review]

**Comment by r4victor** [comment]
Made some adjustments, merging
